### PR TITLE
Align pixel orientation with exif:Orientation in HEIC decoder

### PR DESCRIPTION
**Pull request by rudchenkos**
It turns out that in HEIC the pixels are never rotated and the metadata only
indicates the orientation of the sensor when the photo was taken.

Hence, there was a discrepancy between `Image->orientation` and actual pixel
orientation in images loaded with the `heic` coder. Particularly, this led to
incorrect behavior of the `-auto-orient` which in this case tried to compensate
non-existing pixel rotation.

This is a fix for https://github.com/ImageMagick/ImageMagick/issues/1232

The fix provides two solutions for the problem described above: reset `exif:Orientation` to "1" in HEIC decoder output (default behavior) or rotate pixels to _match_ the original `exif:Orientation`, as some application might rely on the original sensor rotation. The latter is enabled by `heic:preserve-orientation` read define.

### CODE REVIEW

**Comment by rudchenkos** [comment]
A question to the maintainers: where in the docs is the best place to describe the new option `heic:preserve-orientation`?

**Comment by dlemstra** [comment]
Is it not enough to call `SetImageProperty(image,"exif:orientation","1",exception);` after the image has been read? The exif profile will be updated before the image is written with the new value.

**Comment by rudchenkos** [comment]
@dlemstra Absolutely. In most common cases it is really enough to reset `exif:Orientation` after the image is read. This is is the default behavior after my patch.

Unfortunately, there are application which rely on the original pixel rotation for historical reasons. This is why I have also implemented the option `heic:preserve-orientation`.

File: coders/heic.c [hunk 1]
```diff
@@ -85,6 +86,8 @@ static MagickBooleanType
   WriteHEICImage(const ImageInfo *,Image *,ExceptionInfo *);
 #endif
 
+static Image *CompensateOrientation(Image *image, ExceptionInfo *exception);
```
Comment (dlemstra) [inline coders/heic.c]: Could you move this method above the `ReadHEICImage` method, near `IsHeifSuccess`? We can then remove this line.

File: coders/heic.c [hunk 2]
```diff
@@ -299,8 +305,58 @@ static Image *ReadHEICImage(const ImageInfo *image_info,
...
   heif_context_free(heif_context);
+
+  /*
+    There is a discrepancy between EXIF data and the actual orientation of
+    image pixels. ReadImage processes "exif:Orientation" expecting pixels to be
+    oriented accordingly. However, in HEIF the pixels are NOT rotated.
+   */
+  option=GetImageOption(image_info,"heic:preserve-orientation");
+  if (IsStringTrue(option) == MagickTrue)
+    image = CompensateOrientation(image, exception);
```
Comment (dlemstra) [inline coders/heic.c]: This statement leaks memory. The methods `FlipImage`, `RotateImage` return a new image.

File: coders/heic.c [hunk 3]
```diff
@@ -299,8 +305,58 @@ static Image *ReadHEICImage(const ImageInfo *image_info,
...
+  else
+    SetImageProperty(image, "exif:Orientation", "1", exception);
+
   return(GetFirstImageInList(image));
 }
+
+/* An inverse of AutoOrientImage */
+static Image *CompensateOrientation(Image *image, ExceptionInfo *exception)
```
Comment (dlemstra) [inline coders/heic.c]: `Image *image` should be `const Image *image`.

File: coders/heic.c (hunk 1 above)
Comment (rudchenkos) [reply coders/heic.c]: Of course, will do 👍

File: coders/heic.c (hunk 2 above)
Comment (rudchenkos) [reply coders/heic.c]: Oh, my. That's right, I will free the old image here

File: coders/heic.c (hunk 3 above)
Comment (rudchenkos) [reply coders/heic.c]: Makes sense, I will change it to a const pointer type

Comment (dlemstra) [reply coders/heic.c]: Good point. I missed that, probably better to not change the signature to a const.

**Comment by rudchenkos** [comment]
Hi @dlemstra,
I have applied the suggestions in the fixup commit (to be squashed).

About the docs, it seems the most appropriate place to document `heic:preserve-orientation` is [command-line-options.html](https://github.com/ImageMagick/ImageMagick/blob/master/www/command-line-options.html) right after `hough-lines:accumulator` description. One thing is confusing me though, that line (`www/command-line-options.html:84`) is **very** long which makes me think it is generated by a tool, rather than written manually. But, from the other hand I cannot find where from and how could it have been generated.

Should I just edit that `command-line-options.html`?

**Comment by dlemstra** [comment]
You should edit the command line options here: https://github.com/ImageMagick/Website

File: coders/heic.c [hunk 4]
```diff
@@ -299,6 +341,26 @@ static Image *ReadHEICImage(const ImageInfo *image_info,
...
+    There are two solutions to this problem: either reset the EXIF Orientation tag
+    so it matches the orientation of pixels, or rotate the pixels to match EXIF data.
+   */
+  option=GetImageOption(image_info,"heic:preserve-orientation");
+  if (IsStringTrue(option) == MagickTrue)
+    {
+      Image
+        *original_image = image;
+      image = CompensateOrientation(original_image, exception);
+      DestroyImageList(original_image);
```
Comment (dlemstra) [inline coders/heic.c]: You should handle this inside the `CompensateOrientation` method. You are now destroying the `image` when `exif:Orientation` is not set.

Comment (rudchenkos) [reply coders/heic.c]: Oh, true. Fill fix promptly.

Comment (rudchenkos) [reply coders/heic.c]: Pushed.
I am sorry for the silly mistake

**Comment by rudchenkos** [comment]
Pull request for docs: https://github.com/ImageMagick/Website/pull/19

**Comment by rudchenkos** [comment]
Dank u wel, Dirk!

**Pull request merged**